### ISSUE TITLE
Large versioned file sizes

### ISSUE THREAD

**Issue by ArvidJB** [deshaw/versioned-hdf5#192]
We observed that our versioned files become quite large. I included some reproducer code below. Here the difference between the versioned file
```
After 10000 versions final file size is 2672131160
```
and the same code without versioning
```
After 10000 versions final file size is 2528752
```
is roughly 1000 times.

Are we doing something wrong here? Tweaking the chunk size seems to make a difference: smaller chunk sizes lead to smaller sizes. But smaller chunk sizes unfortunately seem to have a big performance penalty? I think it would be great to have some kind of analysis of file sizes vs. performance for different chunk sizes.

```
"""
The versioned file seems to become pretty big compared to the unversioned file.
"""

from __future__ import absolute_import, division

import logging
import os
import shutil
from tempfile import mkdtemp

import h5py
import numpy as np

if __name__ == '__main__':
    logging.getLogger().setLevel(logging.INFO)
    temp_dir = mkdtemp()
    file_name = temp_dir + '/data.h5'
    spns = np.sort(np.random.choice(100000, 1000, replace=False))
    spn_chunk_size = 10
    try:
        with h5py.File(file_name, 'w') as f:
            dates = np.array([20210101], dtype='int32')
            f.create_dataset('date', data=dates, chunks=(10000,), maxshape=(None,))
            f.create_dataset('spn', data=spns, chunks=(10000,), maxshape=(None,))
            values = np.full((len(dates), len(spns)), 0.0, dtype='float')
            f.create_dataset('value', data=values, chunks=(300, 100), maxshape=(None, None))

        v = 1
        for date in range(20210102, 20210201):  # ignore the wrong "dates"
            # add a new "date"
            with h5py.File(file_name, 'r+') as f:
                dates_ds = f['date']
                dates_ds.resize((dates_ds.shape[0] + 1,))
                dates_ds[-1:] = [date]
                values_ds = f['value']
                values_ds.resize((values_ds.shape[0] + 1, values_ds.shape[1]))
                values_ds[-1:, :] = v
            v += 1

            # update values for "spn" in chunks
            spn_chunks = np.array_split(spns, len(spns) // spn_chunk_size)
            for spn_chunk in spn_chunks:
                with h5py.File(file_name, 'r+') as f:
                    spns_ds = f['spn']
                    for spn in spn_chunk:
                        spn_idx = np.searchsorted(spns_ds, spn)
                        if spn_idx < len(spns_ds) and spns_ds[spn_idx] == spn:
                            # update existing spn
                            values_ds = f['value']
                            values_ds[:, spn_idx] = v
                        else:
                            raise NotImplementedError('insert new spn')
                v += 1
    finally:
        print('After %d versions final file size is %d' % (v, os.stat(file_name).st_size,))
        shutil.rmtree(temp_dir)
```

**Comment by asmeurer** [comment]
Your reproducer isn't using Versioned HDF5. I tried adding a `stage_version(str(v))` under each `with File(...)` in your script, but the resulting versioned file size is 17112088, which is only 6 ties bigger than the unversioned file. Did you use a different chunk size when you got 2672131160?

**Comment by ArvidJB** [comment]
Apologies, I posted the unversioned version of the script. This is what I ran
```
"""
The versioned file seems to become pretty big compared to the unversioned file.
"""

from __future__ import absolute_import, division

import logging
import os
import shutil
from tempfile import mkdtemp

import h5py
import numpy as np

from versioned_hdf5 import VersionedHDF5File

if __name__ == '__main__':
    logging.getLogger().setLevel(logging.INFO)
    temp_dir = mkdtemp()
    file_name = temp_dir + '/data.h5'
    spns = np.sort(np.random.choice(100000, 1000, replace=False))
    spn_chunk_size = 10
    try:
        with h5py.File(file_name, 'w') as f:
            vf = VersionedHDF5File(f)
            with vf.stage_version('0') as sv:
                dates = np.array([20210101], dtype='int32')
                sv.create_dataset('date', data=dates, chunks=(10000,), maxshape=(None,))
                sv.create_dataset('spn', data=spns, chunks=(10000,), maxshape=(None,))
                values = np.full((len(dates), len(spns)), 0.0, dtype='float')
                sv.create_dataset('value', data=values, chunks=(300, 100), maxshape=(None, None))

        v = 1
        for date in range(20210102, 20210201):  # ignore the wrong "dates"
            # add a new "date"
            with h5py.File(file_name, 'r+') as f:
                vf = VersionedHDF5File(f)
                with vf.stage_version(str(v)) as sv:
                    dates_ds = sv['date']
                    dates_ds.resize((dates_ds.shape[0] + 1,))
                    dates_ds[-1:] = [date]
                    values_ds = sv['value']
                    values_ds.resize((values_ds.shape[0] + 1, values_ds.shape[1]))
                    values_ds[-1:, :] = v
            v += 1

            # update values for "spn" in chunks
            spn_chunks = np.array_split(spns, len(spns) // spn_chunk_size)
            for spn_chunk in spn_chunks:
                with h5py.File(file_name, 'r+') as f:
                    vf = VersionedHDF5File(f)
                    with vf.stage_version(str(v)) as sv:
                        spns_ds = sv['spn']
                        for spn in spn_chunk:
                            spn_idx = np.searchsorted(spns_ds, spn)
                            if spn_idx < len(spns_ds) and spns_ds[spn_idx] == spn:
                                # update existing spn
                                values_ds = sv['value']
                                values_ds[:, spn_idx] = v
                            else:
                                raise NotImplementedError('insert new spn')
                v += 1
    finally:
        print('After %d versions final file size is %d' % (v, os.stat(file_name).st_size,))
        shutil.rmtree(temp_dir)
```

**Comment by ArvidJB** [comment]
> Did you use a different chunk size when you got 2672131160?

I used `(300, 100)` for both runs. That's sort of what we default to since there are about 300 business dates in one year.

**Comment by ArvidJB** [comment]
After thinking about this a bit more I think the factor 1000 overhead is correct. With a chunk size of `(300, 100)` all "dates" written fall into the same chunk. Since there are 1000 "spns" there are thus only really 10 chunks. Each write will most likely change all 10 chunks such that they contain new "raw data". So each write/version should store a complete new copy.

**Comment by asmeurer** [comment]
I had a comment written up with the same thing, but I never clicked submit. You are only computing 100 days, so each chunk is less than 1/3 full. I did a test with 3000 days (I also reduced spns down to 100) and the resulting ratio of file sizes was `14715240840/2768752 = 5314` for 33000 versions.

**Comment by ArvidJB** [comment]
I think this shows again that we need to pick our `chunks` very carefully. Picking smaller chunk sizes seems like it would improve the file size problem, but then the read/write performance (I will open another issue with an example) suffers. Maybe it is not possible to have both
* small file size overhead for each version
* good read/write performance

**Comment by asmeurer** [comment]
Yes. It is also now possible to adjust the chunk size with the replaying functionality, so you should be able to start with a smaller chunk size and migrate to a larger one once the chunks get full.

Aside from modifying the chunk size, I don't think there's much we can do, at least on the versioned HDF5 side, for file space. But performance is something we do have control over, so we should try to focus on that, IMO. 

Another variable to consider is compression. As I understand it, it works on at the chunk level, so the chunk size may make a difference. I need to test it, but it's also possible compression can take care of the problems with mostly empty chunks, which are filled with the fillvalue in the raw dataset.

**Comment by asmeurer** [comment]
I tried modifying the chunk size of `value` to `(30, 100)`, and it takes the file size down to 513125888 (19% of the original 2677547120). But I also tried adding `compression='gzip'` to `value` (with the `(300, 100)` chunk size), and the file size goes down to 83838847 (3% of the original). Both have some performance penalty, although the compression performance penalty isn't that much, at least for what the test script is doing. Of course, real world data likely won't be as compressible as test data, but I expect a large fraction of those savings are from compressing the empty part of the empty chunks. I didn't play with other compression algorithms or parameters.
